onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc

```diff
@@ -10,101 +10,145 @@
+  Node* first_node_ptr = graph.GetNode(first_node_idx);
+  if (!first_node_ptr) {
+    return false;
+  }
+  Node& first_node = *first_node_ptr;
 
-  for (auto node_index : node_topology_list) {
-    auto* node_ptr = graph.GetNode(node_index);
-    if (node_ptr == nullptr)
-      continue;  // node was removed as part of an earlier optimization
+  const auto match_first = node_sequence_type == NodeSequence::Q_DQ ? &QDQ::MatchQNode : &QDQ::MatchDQNode;
+  const auto match_second = node_sequence_type == NodeSequence::Q_DQ ? &QDQ::MatchDQNode : &QDQ::MatchQNode;
+
+  if (!match_first(first_node) ||
+      // not filtering on provider currently
+      // !graph_utils::IsSupportedProvider(first_node, compatible_execution_providers) ||
+      !optimizer_utils::CheckOutputEdges(graph, first_node, 1)) {
+    return false;
+  }
 
-    Node& q_node = *node_ptr;
-    ORT_RETURN_IF_ERROR(Recurse(q_node, modified, graph_level, logger));
+  Node& second_node = *graph.GetNode(first_node.OutputNodesBegin()->Index());
+  if (!match_second(second_node)
+      // not filtering on provider currently
+      // || !graph_utils::IsSupportedProvider(second_node, compatible_execution_providers)
+  ) {
+    return false;
+  }
 
-    if (!QDQ::MatchQNode(q_node) ||
-        // not filtering on provider currently
-        // !graph_utils::IsSupportedProvider(q_node, GetCompatibleExecutionProviders()) ||
-        !optimizer_utils::CheckOutputEdges(graph, q_node, 1)) {
-      continue;
-    }
+  if (node_sequence_type == NodeSequence::DQ_Q) {
+    // for DQ -> Q, check for constant, matching scale/ZP values
+    const auto get_constant_initializer = [&graph](const std::string& initializer_name) {
+      return graph.GetConstantInitializer(initializer_name, true);
+    };
 
-    Node& dq_node = *graph.GetNode(q_node.OutputNodesBegin()->Index());
-    if (!QDQ::MatchDQNode(dq_node) 
-        // not filtering on provider currently
-        // || !graph_utils::IsSupportedProvider(dq_node, GetCompatibleExecutionProviders())
-        ) {
-      continue;
+    if (!QDQ::IsQDQPairSupported(second_node, first_node, get_constant_initializer, graph.ModelPath())) {
+      return false;
     }
+  }
 
-    // we have Q -> DQ pair
+  // we have a node sequence to clean up
+  LOGS(logger, VERBOSE) << "Cleaning up back-to-back nodes: "
+                        << first_node.OpType() << " with name \"" << first_node.Name() << "\" and "
+                        << second_node.OpType() << " with name \"" << second_node.Name() << "\"";
 
-    // we support a DQ that produces a graph output if it has no output edges, or a DQ node with one output edge.
-    bool is_graph_output = graph.NodeProducesGraphOutput(dq_node);
-    auto edges_count = dq_node.GetOutputEdgesCount();
+  // we support a second_node that produces a graph output if it has no output edges, or a second_node with one output edge.
+  bool is_graph_output = graph.NodeProducesGraphOutput(second_node);
```

Comment:
nit: Is `second_node_produces_graph_output` a name that indicates what is being captured better ?

Reply:
updated name to `produces_graph_output`